bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/BroadlinkSocket.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    @NonNullByDefault
+    private static class ReceiverThread extends Thread {
+        private Logger logger;
+
+        public void run() {
+            receiveData(BroadlinkSocket.socket, BroadlinkSocket.datagramPacket);
+        }
+
+        private void receiveData(@Nullable MulticastSocket socket, DatagramPacket dgram) {
+            try {
+                while (true) {
+                    socket.receive(dgram);
+                    BroadlinkSocketListener listener;
+                    byte remoteMAC[];
+                    org.openhab.core.thing.ThingTypeUID deviceType;
+                    int model;
+                    for (Iterator<BroadlinkSocketListener> iterator = (new ArrayList<BroadlinkSocketListener>(
+                            BroadlinkSocket.listeners)).iterator(); iterator.hasNext(); listener.onDataReceived(
+                                    dgram.getAddress().getHostAddress(), dgram.getPort(), decodeMAC(remoteMAC),
+                                    deviceType, model)) {
+                        listener = iterator.next();
+                        byte receivedPacket[] = dgram.getData();
+                        remoteMAC = Arrays.copyOfRange(receivedPacket, 58, 64);
+                        model = Byte.toUnsignedInt(receivedPacket[52]) | Byte.toUnsignedInt(receivedPacket[53]) << 8;
+                        deviceType = ModelMapper.getThingType(model, logger);
+                    }
+                }
+            } catch (IOException e) {
+                if (!isInterrupted()) {
+                    logger.warn("Error while receiving", e);
+                }
+            }
+            logger.info("Receiver thread ended");
+        }
+
+        private ReceiverThread(Logger logger) {
+            this.logger = logger;
+        }
+    }
+
+    public static void registerListener(BroadlinkSocketListener listener, Logger logger) {
+        listeners.add(listener);
+        if (socket == null) {
+            setupSocket(logger);
+        }
+    }
+
+    public static void unregisterListener(BroadlinkSocketListener listener, Logger logger) {
+        listeners.remove(listener);
+        if (listeners.isEmpty() && socket != null) {
+            closeSocket(logger);
+        }
+    }
+
+    private static void setupSocket(Logger logger) {
+        synchronized (BroadlinkSocket.class) {
+            try {
+                socket = new MulticastSocket();
+            } catch (IOException e) {
+                logger.warn("Setup socket error '{}'.", e.getMessage());
+            }
+            socketReceiveThread = new ReceiverThread(logger);
+            socketReceiveThread.start();
+        }
+    }
+
+    private static void closeSocket(Logger logger) {
+        synchronized (BroadlinkSocket.class) {
+            if (socketReceiveThread != null) {
+                socketReceiveThread.interrupt();
+            }
+            if (socket != null) {
+                logger.info("Socket closed");
```

Comment:
Debug seems a better fit. If this happens a lot, then consider removing this to improve performance. There is a cost each time Java has to check if the logger is set to debug/info etc...